### ISSUE TITLE
Assertions inside implications (sometimes) don't take into account the premise

### ISSUE THREAD

**Issue by jad-hamza** [epfl-lara/stainless#425]
```scala
import stainless.lang._

case class T(var x: BigInt) {
  def f() = {
    x = x + 1 // commenting this makes the assertion go through
    false ==> { assert(false); true }
  }
}
```

```scala
[  Info  ]   ┌───────────────────┐
[  Info  ] ╔═╡ stainless summary ╞════════════════════════════════════════════════
[  Info  ] ║ └───────────────────┘                                              
[  Info  ] ║ f  body assertion    invalid   nativez3   File.scala:6:15      0,026 
[  Info  ] ╟┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
[  Info  ] ║ total: 1    valid: 0    (0 from cache) invalid: 1    unknown: 0    time:   0,026
[  Info  ] ╚══════════════════════════════════════════════════════════════════════
```

(When we comment the variable assignment the assertion goes through)

**Comment by romac** [comment]
After `ImperativeCodeElimination`:

### With `x = x + 1` (invalid)

```scala
[info] [ Debug  ] def f(thiss: T): (Boolean, T) = {
[info] [ Debug  ]   val thiss: T = thiss
[info] [ Debug  ]   val thiss: T = T({
[info] [ Debug  ]     val e: T = thiss
[info] [ Debug  ]     @unchecked e.x
[info] [ Debug  ]   } + 1)
[info] [ Debug  ]   assert(false)
[info] [ Debug  ]   (false ==> true, thiss)
[info] [ Debug  ] }
```

### Without `x = x + 1` (valid)

```scala
[info] [ Debug  ] def f(thiss: T): Boolean = false ==> {
[info] [ Debug  ]   assert(false)
[info] [ Debug  ]   true
[info] [ Debug  ] }
```

The issue seems to stem from https://github.com/epfl-lara/stainless/blob/93dbd338e1814994aa273da62f33a7f212c7610e/core/src/main/scala/stainless/extraction/imperative/ImperativeCodeElimination.scala#L365-L371, which lifts the assertion outside of the implication. Not sure what do about this, as I don't know what kind of bug this is supposed to work around. @samarion Any pointers?

**Comment by romac** [comment]
Fixed in 5be7475f251a158e967723706f57e13fd50ba69e

**Comment by romac** [comment]
Fixed in 73c4882
